hssqlppp/src/Database/HsSqlPpp/Parsing/ParserInternal.lhs

```diff
@@ -883,19 +885,68 @@ ddl
 >      return $ CreateTable p tname atts cons pdata rep ts
 >     ]
 >   where
->     --parse the unordered list of attribute defs or constraints, for
->     --each line want to try the constraint parser first, then the
->     --attribute parser, so you need the swap to feed them in the
->     --right order into createtable
->     readAttsAndCons =
->               parens (swap <$> multiPerm
->                                  (try tableConstraint)
->                                  tableAttribute
->                                  (symbol ","))
->               where swap (a,b) = (b,a)
 >     readPartition = tryOptionMaybe (tablePartition)
 
->
+> creatExternalTable :: SParser Statement
+> creatExternalTable = do
+>   p <- pos
+>   rep <- choice
+>     [ NoReplace <$ mapM_ keyword ["external", "table"]
+>     , Replace <$ mapM_ keyword ["or", "replace", "external", "table"]
+>     ]
+>   etname <- name
+>   (atts, constraints) <- readAttsAndCons
+>   unless (null constraints)
+>     $ fail $ unlines
+>       [ "CREATE EXTERNAL TABLE does not support table constraints."
+>       ]
+>   keyword "using"
+>   format <- do
+>     keyword "format"
+>     let
+>       formats =
+>         [ (ETFCsv, "csv")
+>         , (ETFParquet, "parquet")
+>         ]
+>     choice $
+>       map (\(constructor, format) -> constructor <$ keyword format) formats
+>       ++
+>         [ fail $ "Unexpected external table format. Expecting one of: "
+>           ++ intercalate ", " (map snd formats)
+>         ]
+>   keyword "with"
+>   opts <-
+>     case format of
+>       ETFCsv -> do
+>         csvOpts <- externalOptionsCsv
```

Comment:
This can be more easily written like this: `EtCsvOptions <$> externalOptionsCsv`